src/reader/mod.rs

```diff
@@ -195,41 +204,59 @@ impl InnerIndexReader {
-        &self,
         segment_readers: &[SegmentReader],
+        searcher_generation_counter: &Arc<AtomicU64>,
+        searcher_generation_inventory: &Inventory<SearcherGeneration>,
     ) -> TrackedObject<SearcherGeneration> {
-        let generation_id = self
-            .searcher_generation_counter
-            .fetch_add(1, atomic::Ordering::Relaxed);
+        let generation_id = searcher_generation_counter.fetch_add(1, atomic::Ordering::Relaxed);
         let searcher_generation =
             SearcherGeneration::from_segment_readers(segment_readers, generation_id);
-        self.searcher_generation_inventory
-            .track(searcher_generation)
+        searcher_generation_inventory.track(searcher_generation)
+    }
+
+    fn create_searcher(
+        index: &Index,
+        doc_store_cache_size: usize,
+        warming_state: &WarmingState,
+        searcher_generation: TrackedObject<SearcherGeneration>,
+    ) -> crate::Result<Searcher> {
+        let segment_readers = Self::open_segment_readers(index)?;
+        let schema = index.schema();
+        let searcher: Searcher = Searcher::new(
+            schema,
+            index.clone(),
+            segment_readers,
+            searcher_generation,
+            doc_store_cache_size,
+        )?;
+        warming_state.warm_new_searcher_generation(&searcher)?;
+        Ok(searcher)
     }
 
     fn reload(&self) -> crate::Result<()> {
-        let segment_readers = self.open_segment_readers()?;
-        let searcher_generation = self.create_new_searcher_generation(&segment_readers);
-        let schema = self.index.schema();
-        let searchers: Vec<Searcher> = std::iter::repeat_with(|| {
-            Searcher::new(
-                schema.clone(),
-                self.index.clone(),
-                segment_readers.clone(),
-                searcher_generation.clone(),
-                self.doc_store_cache_size,
-            )
-        })
-        .take(self.num_searchers)
-        .collect::<io::Result<_>>()?;
-        self.warming_state
-            .warm_new_searcher_generation(&searchers[0])?;
-        self.searcher_pool.publish_new_generation(searchers);
+        let segment_readers = Self::open_segment_readers(&self.index)?;
+        let searcher_generation = Self::create_new_searcher_generation(
+            &segment_readers,
+            &self.searcher_generation_counter,
+            &self.searcher_generation_inventory,
+        );
+        let searcher = Self::create_searcher(
+            &self.index,
+            self.doc_store_cache_size,
+            &self.warming_state,
+            searcher_generation,
+        )?;
+
+        {
+            let mut old_searcher = self.searcher.lock().unwrap();
+            *old_searcher = searcher;
+        }
+
         Ok(())
     }
 
-    fn searcher(&self) -> LeasedItem<Searcher> {
-        self.searcher_pool.acquire()
+    fn searcher(&self) -> Searcher {
+        self.searcher.lock().unwrap().clone()
```

Comment:
The lock could be a problem for high QPS use cases.
(we recommend fetching a new searcher for every request).

Could you check if crossbeam::AtomicCell are lockless and perf well?
Alternatively ArcSwap but i htink the latter is not part of the current deps.

Reply:
`AtomicCell::load()` only works for `Copy` types, so `ArcSwap` is probably the way to go!

(a naive question from me here: https://github.com/vorner/arc-swap/issues/73)

Reply:
Ah ah :) it's a small world.